pip_connectors/pip-connector-seachange-common/src/pip/connector/seachange/common/utilities.py

```diff
@@ -0,0 +1,184 @@
+
+    Args:
+        host_url (str): IOTICSpace (Host) url
+
+    Returns:
+        dict: info to connect to the Host.
+    """
+
+    if not host_url:
+        log.error("Parameter HOST_URL not set")
+        sys.exit(1)
+
+    index_json: str = host_url + constant.INDEX_JSON_PATH
+    req_resp: dict = {}
+
+    try:
+        req_resp = requests.get(index_json, timeout=3).json()
+    except requests.exceptions.ConnectionError:
+        log.error("Can't connect to %s. Check HOST_URL is spelt correctly", index_json)
+        sys.exit(1)
+
+    return req_resp
+
+
+def auto_refresh_token(
+    refresh_token_lock: Lock, identity: Identity, iotics_api: IoticsApi
+):
+    """Automatically refresh then IOTICS token before it expires.
+
+    Args:
+        refresh_token_lock (Lock): used to prevent race conditions.
+        identity (Identity): the instance of Identity API used to manage IOTICS Identities.
+        iotics_api (IoticsApi): the instance of IOTICS gRPC API used to execute Twins operations.
+    """
+
+    token_period = int(identity.token_duration * constant.TOKEN_REFRESH_PERIOD_PERCENT)
+
+    while True:
+        time_to_refresh: int = token_period - (time() - identity.token_last_updated)
+        sleep(time_to_refresh)
+        with refresh_token_lock:
+            identity.refresh_token()
+            iotics_api.update_channel()
+
+        log.debug("Token refreshed correctly")
+
+
+def search_twins(
+    search_criteria: search_pb2.SearchRequest.Payload,
+    refresh_token_lock: Lock,
+    iotics_api: IoticsApi,
+    keep_searching: bool = True,
+):
+    """Wrapper of the Search Twin operation to help checking for errors
+    and retrying when they occur.
+
+    Args:
+        search_criteria (Payload): search criteria for Twins in terms of
+            text, properties and/or location;
+        refresh_token_lock (Lock): used to prevent race conditions.
+        iotics_api (IoticsApi): the instance of Identity API used to manage IOTICS Identities
+        keep_searching (bool, optional): whether to keep searching if the result is an empty list.
+            Defaults to True.
+
+    Returns:
+        twins_found_list: list of Twins found.
+    """
+
+    twins_found_list = []
+
+    while True:
+        for attempt in range(constant.RETRYING_ATTEMPTS):
+            try:
+                with refresh_token_lock:
+                    for response in iotics_api.search_iter(
+                        client_app_id=uuid4().hex, payload=search_criteria
+                    ):
+                        twins = response.payload.twins
+                        twins_found_list.extend(twins)
+            except grpc.RpcError as ex:
+                if not expected_grpc_exception(exception=ex, operation="search_twins"):
+                    break
+                log.debug("Attempt #%d", attempt + 1)
+            else:
+                break
+
+        if not twins_found_list and keep_searching:
```

Comment:
The "attempt in range" loop is more about retrying in case of grpc errors.
The `if not twins_found_list and keep_searching:` block instead can be useful if, for instance, the Twins to search for haven't been created yet, but will be created soon

Reply:
but the retry loop has no retry delay? 
I would say either move this block in for both cases or add an additional retry delay?

Reply:
There's a `sleep` if a new iteration is needed ([here](https://github.com/Iotic-Labs/iotics-tutorials-py/blob/dev-lp-vessel-movement-connector/pip_connectors/pip-connector-seachange-common/src/pip/connector/seachange/common/utilities.py#L111))

Reply:
Sorry I understand the first block has no delay. I'll change it